Parse the Azure OpenAI rate limit message

tombl opened the pull request:
Fixes #4161

Currently Codex uses a regex to parse the "Please try again in 1.898s" OpenAI-style rate limit message, so that it can wait the correct duration before retrying. Azure OpenAI returns a different error that looks like "Rate limit exceeded. Try again in 35 seconds."

This PR extends the regex and parsing code to match in a more fuzzy manner, handling anything matching the pattern "try again in \<duration>\<unit>".

tombl committed:
add failing test

tombl committed:
implement

etraut-openai commented:
@codex review

etraut-openai commented on codex-rs/core/src/client.rs:
```diff
@@ -931,32 +931,44 @@ async fn stream_from_fixture(
 fn rate_limit_regex() -> &'static Regex {
     static RE: OnceLock<Regex> = OnceLock::new();
 
+    // Match both the existing OpenAI-style "Please try again in 1.898s"
```
This comment looks like it was generated by AI. Please update so it's clear that the Regex is intended to match the response from both OpenAI and Azure response endpoints. Provide examples of both formats. This will help avoid regressions in the future.

tombl replied:
Updated

etraut-openai commented on codex-rs/core/src/client.rs:
```diff
@@ -931,32 +931,44 @@ async fn stream_from_fixture(
 
-            if unit == "s" {
-                return Some(Duration::from_secs_f64(value));
+            return if unit == "s" || unit.starts_with("second") {
```
Does the Azure format use "second" rather than "s"? If so, this is OK. Otherwise, this seems unnecessary.

tombl commented:
I've updated the PR to be the bare minimum diff necessary.

It was written with LLM assistance but ultimately the choice to defensively handle other units was mine. I've done some more research and it looks like Azure does always return seconds, so I've reverted that part.

(also your response wasn't slow at all, I'm used to corporate open source projects taking much longer to review community PRs, so thank you!)
